File: perception-core/src/main/java/com/hh/uiperception/core/semantic/SemanticNode.java
package com.hh.uiperception.core.semantic;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 面向模型消费的 native 语义节点。
 */
public final class SemanticNode {

    private static final int NO_WEB_ELEMENT = -1;

    private final SemanticRole role;
    private final String name;
    private final String text;
    private final String contentDescription;
    private final String resourceId;
    private final String className;
    private final Bounds bounds;
    private final String ref;
    private final RoleDecision roleDecision;
    private final List<String> states;
    private final List<SemanticNode> children;
    private final int webElementIdx;

    private SemanticNode(Builder builder) {
        if (builder.role == null) {
            throw new IllegalArgumentException("role must not be null");
        }
        this.role = builder.role;
        this.name = normalize(builder.name);
        this.text = normalize(builder.text);
        this.contentDescription = normalize(builder.contentDescription);
        this.resourceId = normalize(builder.resourceId);
        this.className = normalize(builder.className);
        this.bounds = builder.bounds;
        this.ref = normalize(builder.ref);
        this.roleDecision = builder.roleDecision;
        this.states = Collections.unmodifiableList(new ArrayList<>(builder.states));
        this.children = Collections.unmodifiableList(new ArrayList<>(builder.children));
        this.webElementIdx = builder.webElementIdx;
    }

    public static Builder builder(SemanticRole role) {
        return new Builder(role);
    }

    public SemanticRole role() {
        return role;
    }

    public String name() {
        return name;
    }

    public String text() {
        return text;
    }

    public String contentDescription() {
        return contentDescription;
    }

    public String resourceId() {
        return resourceId;
    }

    public String className() {
        return className;
    }

    public Bounds bounds() {
        return bounds;
    }

    public String ref() {
        return ref;
    }

    public RoleDecision roleDecision() {
        return roleDecision;
    }

    public List<String> states() {
        return states;
    }

    public List<SemanticNode> children() {
        return children;
    }

    public boolean hasRef() {
        return !ref.isEmpty();
    }

    public int webElementIdx() {
        return webElementIdx;
    }

    private static String normalize(String value) {
        if (value == null) {
            return "";
        }
        return value.trim().replaceAll("\\s+", " ");
    }

    public static final class Builder {
        private final SemanticRole role;
        private String name;
        private String text;
        private String contentDescription;
        private String resourceId;
        private String className;
        private Bounds bounds;
        private String ref;
        private RoleDecision roleDecision;
        private final List<String> states = new ArrayList<>();
        private final List<SemanticNode> children = new ArrayList<>();
        private int webElementIdx = NO_WEB_ELEMENT;

        private Builder(SemanticRole role) {
            this.role = role;
        }

        public Builder name(String name) {
            this.name = name;
            return this;
        }

        public Builder text(String text) {
            this.text = text;
            return this;
        }

        public Builder contentDescription(String contentDescription) {
            this.contentDescription = contentDescription;
            return this;
        }

        public Builder resourceId(String resourceId) {
            this.resourceId = resourceId;
            return this;
        }

        public Builder className(String className) {
            this.className = className;
            return this;
        }

        public Builder bounds(Bounds bounds) {
            this.bounds = bounds;
            return this;
        }

        public Builder ref(String ref) {
            this.ref = ref;
            return this;
        }

        public Builder roleDecision(RoleDecision roleDecision) {
            this.roleDecision = roleDecision;
            return this;
        }

        public Builder addState(String state) {
            String normalized = normalize(state);
            if (!normalized.isEmpty()) {
                states.add(normalized);
            }
            return this;
        }

        public Builder addChild(SemanticNode child) {
            if (child != null) {
                children.add(child);
            }
            return this;
        }

        public Builder webElementIdx(int webElementIdx) {
            this.webElementIdx = webElementIdx;
            return this;
        }

        public SemanticNode build() {
            return new SemanticNode(this);
        }
    }
}

File: ios/UIPerception/Capture/SemanticNode.swift
import Foundation

/// 面向模型消费的 native 语义节点，对应 Android `SemanticNode.java`。
final class SemanticNode {
    let role: SemanticRole
    let name: String
    let text: String
    let contentDescription: String
    let resourceId: String
    let className: String
    let bounds: Bounds?
    let ref: String
    let roleDecision: RoleDecision?
    let states: [String]
    let children: [SemanticNode]
    let webElementIdx: Int

    static let noWebElement: Int = -1

    private init(role: SemanticRole,
                 name: String?,
                 text: String?,
                 contentDescription: String?,
                 resourceId: String?,
                 className: String?,
                 bounds: Bounds?,
                 ref: String?,
                 roleDecision: RoleDecision?,
                 states: [String],
                 children: [SemanticNode],
                 webElementIdx: Int) {
        self.role = role
        self.name = SemanticNode.normalize(name)
        self.text = SemanticNode.normalize(text)
        self.contentDescription = SemanticNode.normalize(contentDescription)
        self.resourceId = SemanticNode.normalize(resourceId)
        self.className = SemanticNode.normalize(className)
        self.bounds = bounds
        self.ref = SemanticNode.normalize(ref)
        self.roleDecision = roleDecision
        self.states = states
        self.children = children
        self.webElementIdx = webElementIdx
    }

    var hasRef: Bool { !ref.isEmpty }

    static func builder(_ role: SemanticRole) -> Builder { Builder(role) }

    private static func normalize(_ value: String?) -> String {
        guard let value else { return "" }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let collapsed = trimmed.replacingOccurrences(
            of: "\\s+",
            with: " ",
            options: .regularExpression
        )
        return collapsed
    }

    final class Builder {
        private let role: SemanticRole
        private var name: String?
        private var text: String?
        private var contentDescription: String?
        private var resourceId: String?
        private var className: String?
        private var bounds: Bounds?
        private var ref: String?
        private var roleDecision: RoleDecision?
        private var states: [String] = []
        private var children: [SemanticNode] = []
        private var webElementIdx: Int = SemanticNode.noWebElement

        init(_ role: SemanticRole) { self.role = role }

        @discardableResult
        func name(_ v: String?) -> Builder { self.name = v; return self }

        @discardableResult
        func text(_ v: String?) -> Builder { self.text = v; return self }

        @discardableResult
        func contentDescription(_ v: String?) -> Builder { self.contentDescription = v; return self }

        @discardableResult
        func resourceId(_ v: String?) -> Builder { self.resourceId = v; return self }

        @discardableResult
        func className(_ v: String?) -> Builder { self.className = v; return self }

        @discardableResult
        func bounds(_ v: Bounds?) -> Builder { self.bounds = v; return self }

        @discardableResult
        func ref(_ v: String?) -> Builder { self.ref = v; return self }

        @discardableResult
        func roleDecision(_ v: RoleDecision?) -> Builder { self.roleDecision = v; return self }

        @discardableResult
        func addState(_ v: String?) -> Builder {
            let normalized = SemanticNode.normalize(v)
            if !normalized.isEmpty { states.append(normalized) }
            return self
        }

        @discardableResult
        func addChild(_ child: SemanticNode?) -> Builder {
            if let child { children.append(child) }
            return self
        }

        @discardableResult
        func webElementIdx(_ v: Int) -> Builder { self.webElementIdx = v; return self }

        func build() -> SemanticNode {
            SemanticNode(
                role: role,
                name: name,
                text: text,
                contentDescription: contentDescription,
                resourceId: resourceId,
                className: className,
                bounds: bounds,
                ref: ref,
                roleDecision: roleDecision,
                states: states,
                children: children,
                webElementIdx: webElementIdx
            )
        }
    }
}
